### PR TITLE
BugsnagUnity v4

### PR DESCRIPTION
**Pull request by martin308**
## Goals

- Support windows platforms
- Improve support for all of the features of Bugsnag across all of the platforms

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/Bugsnag.Unity/Payload/App.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    [DllImport("bugsnag-osx", EntryPoint = "retrieveAppData")]
+    static extern void RetrieveAppData(IntPtr instance, Action<IntPtr, string, string> populate);
+
+    internal MacOsApp(IConfiguration configuration) : base(configuration)
+    {
+      GCHandle handle;
+
+      try
+      {
+        handle = GCHandle.Alloc(this);
```
Comment (kattrali) [inline src/Bugsnag.Unity/Payload/App.cs]: This bit doesn't build on windows, as it complains that the `Free()` call is possible invoked on an uninitialized `handle` variable.

I changed it to:

```cs
      GCHandle handle;

      try
      {
        handle = GCHandle.Alloc(this);

        try
        {
          RetrieveAppData(GCHandle.ToIntPtr(handle), PopulateAppData);
        }
        finally
        {
          handle.Free();
        }
      }
      catch (ArgumentException _ex)
      {
        
      }
```

Though there might be a better way of doing this.

File: src/Bugsnag.Unity/Payload/App.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    [DllImport("__Internal", EntryPoint = "retrieveAppData")]
+    static extern void RetrieveAppData(IntPtr instance, Action<IntPtr, string, string> populate);
+
+    internal iOSApp(IConfiguration configuration) : base(configuration)
+    {
+      GCHandle handle;
+
+      try
+      {
+        handle = GCHandle.Alloc(this);
```
Comment (kattrali) [inline src/Bugsnag.Unity/Payload/App.cs]: Same as above

File: src/Bugsnag.Unity/Payload/App.cs [hunk 3]
```diff
@@ -43,17 +43,18 @@ class MacOsApp : App
 
     internal MacOsApp(IConfiguration configuration) : base(configuration)
     {
-      GCHandle handle;
+      GCHandle handle = GCHandle.Alloc(this);
```
[the remark this answers is not in the document]
Comment (martin308) [reply src/Bugsnag.Unity/Payload/App.cs]: From what I have read it could, but then you would have nothing to free

[the remark this answers is not in the document]
Comment (kattrali) [reply src/Bugsnag.Unity/Payload/App.cs]: Ah true, but should we catch the exception?

**Comment by Donnotron666** [comment]
Hi, I've been following this and have even integrated it into my project. I will say that while newer C# features are nice, it breaks compat for game projects that are also exporting to the nintendo switch, which doesn't allow newer .net features. This is a huge headache for something that's mostly code-cosmetic.  Even though projects that target the switch won't be able to use bugsnag, I'd still like to take advantage of it for my windows builds.

**Comment by martin308** [comment]
@Donnotron666 can you be more specific about what errors you are seeing? We are targeting .net35 with this build. I can look into fixing what issues you are seeing if you can let us know more
